sdk/core/Microsoft.Azure.Core.Spatial.NewtonsoftJson/src/JsonExtensions.cs

```diff
@@ -0,0 +1,277 @@
+        public static void ExpectAndAdvance(
+            this JsonReader reader,
+            JsonToken expectedToken,
+            params object[] expectedValues) => ExpectAndAdvance<object>(reader, expectedToken, expectedValues);
+
+        /// <summary>
+        /// Asserts that the given JSON reader is positioned on a token with the expected type and retrieves the
+        /// value of the token, if any. Optionally asserts that the value of the token matches a given expected
+        /// value. If any of the assertions fail, this method throws a JsonSerializationException. Otherwise, this
+        /// method attempts to advance the JSON reader to the next position.
+        /// </summary>
+        /// <typeparam name="TValue">The expected type of the value of the current JSON token.</typeparam>
+        /// <param name="reader">The JSON reader.</param>
+        /// <param name="expectedToken">The JSON token on which the reader is expected to be positioned.</param>
+        /// <param name="expectedValues">Optional; The expected possible values of the current JSON token.</param>
+        /// <returns>
+        /// The value of the JSON token before advancing the reader, or default(TValue) if the token has no value.
+        /// </returns>
+        public static TValue ExpectAndAdvance<TValue>(
+            this JsonReader reader,
+            JsonToken expectedToken,
+            params object[] expectedValues)
+        {
+            TValue result = Expect<TValue>(reader, expectedToken, expectedValues);
+            Advance(reader);
+            return result;
+        }
+
+        /// <summary>
+        /// Asserts that the given JSON reader is positioned on a token with the expected type and retrieves the
+        /// value of the token, if any. Optionally asserts that the value of the token matches a given expected
+        /// value. If any of the assertions fail, this method throws a JsonSerializationException.
+        /// </summary>
+        /// <typeparam name="TValue">The expected type of the value of the current JSON token.</typeparam>
+        /// <param name="reader">The JSON reader.</param>
+        /// <param name="expectedToken">The JSON token on which the reader is expected to be positioned.</param>
+        /// <param name="expectedValues">Optional; The expected possible values of the current JSON token.</param>
+        /// <returns>
+        /// The value of the current JSON token, or default(TValue) if the current token has no value.
+        /// </returns>
+        public static TValue Expect<TValue>(
+            this JsonReader reader,
+            JsonToken expectedToken,
+            params object[] expectedValues)
+        {
+            Argument.AssertNotNull(reader, nameof(reader));
+
+            if (reader.TokenType != expectedToken)
+            {
+                throw new JsonSerializationException(
```

Comment:
Do these capture line/column?  If not, is there any simple overload that could or way of extracting that from the reader?  Might be helpful if something in the guts of a massive payload no longer works quite the same way.

Reply:
From a streaming payload, it would be line 1 column <huge number>. I don't see how that's helpful. That only works for formatted documents.

Reply:
I don't care if it's a line/column, an offset, etc.  It's more about if there's an issue here, does this report where that issue occurred in the JSON when it finally bubbles up to users?  STJ will do that automagically and I wanted to double check Newtonsoft does too.

Reply:
I know from experience it does during deserialization (i.e. reflecting on members, etc., but I suppose much of that is also going through converters at some point). It also has [members](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_JsonSerializationException.htm) but I can fail a test and double check in this specific instance it does.

Reply:
Since this PR doesn't have a complete view of the converters, I'll add tests to make sure errors are actionable in the PR #15468.